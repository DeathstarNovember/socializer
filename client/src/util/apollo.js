import ApolloClient from "apollo-client";
import { InMemoryCache } from "apollo-cache-inmemory";
import * as AbsintheSocket from "@absinthe/socket";
import { createAbsintheSocketLink } from "@absinthe/socket-apollo-link";
import { Socket as PhoenixSocket, LongPoll } from "phoenix";
import { createHttpLink } from "apollo-link-http";
import { hasSubscription } from "@jumpn/utils-graphql";
import { split } from "apollo-link";
import { setContext } from "apollo-link-context";
import Cookies from "js-cookie";

const HTTP_URI =
  process.env.NODE_ENV === "production"
    ? "https://brisk-hospitable-indianelephant.gigalixirapp.com"
    : "http://localhost:4000";

const WS_URI =
  process.env.NODE_ENV === "production"
    ? "wss://brisk-hospitable-indianelephant.gigalixirapp.com/socket"
    : "ws://localhost:4000/socket";

function createPhoenixSocket() {
  // Socket with full fallback to LongPoll
  // via https://elixirforum.com/t/fall-back-to-longpoll-when-websocket-fails/23894
  const socket = new PhoenixSocket(WS_URI, {
    params: () => {
      if (Cookies.get("token")) {
        return { token: Cookies.get("token") };
      } else {
        return {};
      }
    },
  });
  socket.onError(() => {
    if (navigator.onLine) {
      const isWebSocket = socket.transport === window.WebSocket;
      console.error(
        `Error connecting using ${isWebSocket ? "WebSocket" : "LongPoll"}`,
      );
      if (isWebSocket) socket.transport = LongPoll;
      else if (window.WebSocket) socket.transport = window.WebSocket;
    }
  });
  return socket;
}

export const createClient = ({ ssr, req, fetch, tokenCookie } = {}) => {
  let link = createHttpLink({ uri: HTTP_URI, fetch });
  let absintheSocket;

  if (!ssr) {
    absintheSocket = AbsintheSocket.create(createPhoenixSocket());
    const socketLink = createAbsintheSocketLink(absintheSocket);

    link = split(
      (operation) => hasSubscription(operation.query),
      socketLink,
      link,
    );
  }

  const authLink = setContext((_, { headers }) => {
    // Get the authentication token from the cookie if it exists.
    const token = tokenCookie || Cookies.get("token");

    // Return the headers to the context so httpLink can read them.
    return {
      headers: {
        ...headers,
        authorization: token ? `Bearer ${token}` : "",
      },
    };
  });

  const cache = new InMemoryCache();
  if (!ssr) {
    cache.restore(window.__APOLLO_STATE__);
  }

  const client = new ApolloClient({
    connectToDevTools: !ssr,
    ssrMode: ssr,
    link: authLink.concat(link),
    cache,
  });
  return { client, absintheSocket };
};

export const refreshSocket = (socket) => {
  // Close the connection to force a reconnection with the
  // new token parameter.
  socket.phoenixSocket.conn && socket.phoenixSocket.conn.close();
};
