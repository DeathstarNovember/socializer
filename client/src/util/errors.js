import { camelizeKeys } from "humps";

export const errorHash = ({ graphQLErrors } = {}) => {
  if (!Array.isArray(graphQLErrors)) return {};
  return camelizeKeys(
    graphQLErrors.reduce(
      (hash, error) => Object.assign(hash, { [error.field]: error.message }),
      {},
    ),
  );
};
