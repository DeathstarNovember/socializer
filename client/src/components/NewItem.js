import React, { useContext, useState } from "react";
import { useMutation } from "@apollo/react-hooks";
import gql from "graphql-tag";
import { Button, Card, Form } from "react-bootstrap";
import { Link } from "react-router-dom";
import renderIf from "render-if";
import { AuthContext } from "util/context";

export const CREATE_POST = gql`
  mutation CreatePost($body: String!) {
    createPost(body: $body) {
      id
    }
  }
`;

export const CREATE_COMMENT = gql`
  mutation CreateComment($postId: String!, $body: String!) {
    createComment(postId: $postId, body: $body) {
      id
    }
  }
`;

const NewItem = ({ feedType, params }) => {
  const { token } = useContext(AuthContext);
  const [body, setBody] = useState("");

  const mutation = feedType === "comment" ? CREATE_COMMENT : CREATE_POST;

  const [submit, { loading }] = useMutation(mutation, {
    onCompleted: () => setBody(""),
  });

  return (
    <Card className="mb-4">
      <Card.Body>
        {renderIf(token)(
          <Form
            onSubmit={(e) => {
              e.preventDefault();
              submit({ variables: { body, ...params } });
            }}
          >
            <Form.Group>
              <Form.Control
                as="textarea"
                rows="3"
                placeholder="What's on your mind?"
                value={body}
                onChange={(e) => setBody(e.target.value)}
              />
            </Form.Group>

            <Button variant="primary" type="submit" disabled={loading}>
              {loading ? "Submitting..." : "Submit"}
            </Button>
          </Form>,
        )}
        {renderIf(!token)(
          <div className="text-muted h7">
            <Link to="/login">Log in</Link> to submit a {feedType}.
          </div>,
        )}
      </Card.Body>
    </Card>
  );
};

export default NewItem;
