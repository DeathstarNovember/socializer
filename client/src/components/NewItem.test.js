import React from "react";
import { render, fireEvent, wait } from "@testing-library/react";
import { MockedProvider } from "@apollo/react-testing";
import { MemoryRouter } from "react-router-dom";
import { AuthContext } from "util/context";
import NewItem, { CREATE_POST, CREATE_COMMENT } from "./NewItem";

describe("NewItem", () => {
  it("renders correctly when not authenticated", () => {
    const { container } = render(
      <MemoryRouter>
        <AuthContext.Provider value={{}}>
          <MockedProvider mocks={[]} addTypename={false}>
            <NewItem feedType="post" />
          </MockedProvider>
        </AuthContext.Provider>
      </MemoryRouter>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  describe("when creating new post", () => {
    it("renders correctly", () => {
      const { container } = render(
        <MemoryRouter>
          <AuthContext.Provider value={{ token: "abc" }}>
            <MockedProvider mocks={[]} addTypename={false}>
              <NewItem feedType="post" />
            </MockedProvider>
          </AuthContext.Provider>
        </MemoryRouter>,
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it("creates new comment", async () => {
      const mocks = [
        {
          request: {
            query: CREATE_POST,
            variables: { body: "New post" },
          },
          result: { data: { createPost: { id: 123 } } },
        },
      ];
      const { getByPlaceholderText, getByText } = render(
        <MemoryRouter>
          <AuthContext.Provider value={{ token: "abc" }}>
            <MockedProvider mocks={mocks} addTypename={false}>
              <NewItem feedType="post" />
            </MockedProvider>
          </AuthContext.Provider>
        </MemoryRouter>,
      );
      fireEvent.change(getByPlaceholderText("What's on your mind?"), {
        target: { value: "New post" },
      });
      fireEvent.click(getByText("Submit"));
      await wait(() =>
        expect(getByPlaceholderText("What's on your mind?").value).toEqual(""),
      );
    });
  });

  describe("when creating new comment", () => {
    it("renders correctly", () => {
      const { container } = render(
        <MemoryRouter>
          <AuthContext.Provider value={{ token: "abc" }}>
            <MockedProvider mocks={[]} addTypename={false}>
              <NewItem feedType="comment" params={{ postId: 123 }} />
            </MockedProvider>
          </AuthContext.Provider>
        </MemoryRouter>,
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it("creates new comment", async () => {
      const mocks = [
        {
          request: {
            query: CREATE_COMMENT,
            variables: { body: "New comment", postId: 123 },
          },
          result: { data: { createComment: { id: 456 } } },
        },
      ];
      const { getByPlaceholderText, getByText } = render(
        <MemoryRouter>
          <AuthContext.Provider value={{ token: "abc" }}>
            <MockedProvider mocks={mocks} addTypename={false}>
              <NewItem feedType="comment" params={{ postId: 123 }} />
            </MockedProvider>
          </AuthContext.Provider>
        </MemoryRouter>,
      );
      fireEvent.change(getByPlaceholderText("What's on your mind?"), {
        target: { value: "New comment" },
      });
      fireEvent.click(getByText("Submit"));
      await wait(() =>
        expect(getByPlaceholderText("What's on your mind?").value).toEqual(""),
      );
    });
  });
});
