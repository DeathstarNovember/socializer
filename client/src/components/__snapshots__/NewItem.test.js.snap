// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewItem renders correctly when not authenticated 1`] = `
<div
  class="mb-4 card"
>
  <div
    class="card-body"
  >
    <div
      class="text-muted h7"
    >
      <a
        href="/login"
      >
        Log in
      </a>
       to submit a 
      post
      .
    </div>
  </div>
</div>
`;

exports[`NewItem when creating new comment renders correctly 1`] = `
<div
  class="mb-4 card"
>
  <div
    class="card-body"
  >
    <form
      class=""
    >
      <div
        class="form-group"
      >
        <textarea
          class="form-control"
          placeholder="What's on your mind?"
          rows="3"
        />
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;

exports[`NewItem when creating new post renders correctly 1`] = `
<div
  class="mb-4 card"
>
  <div
    class="card-body"
  >
    <form
      class=""
    >
      <div
        class="form-group"
      >
        <textarea
          class="form-control"
          placeholder="What's on your mind?"
          rows="3"
        />
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;
